python/ray/tune/utils/file_transfer.py

```diff
@@ -0,0 +1,236 @@
+            tar.add(source_dir, arcname="", recursive=True)
+        else:
+            # Otherwise, only pack differing files
+            tar.add(source_dir, arcname="", recursive=False)
+            for root, dirs, files in os.walk(source_dir, topdown=False):
+                rel_root = os.path.relpath(root, source_dir)
+                # Always add all directories
+                for dir in dirs:
+                    key = os.path.join(rel_root, dir)
+                    tar.add(os.path.join(source_dir, key), arcname=key, recursive=False)
+                # Add files where our information differs
+                for file in files:
+                    key = os.path.join(rel_root, file)
+                    stat = os.lstat(os.path.join(source_dir, key))
+                    file_stat = stat.st_mtime, stat.st_size
+                    if key not in files_stats or file_stat != files_stats[key]:
+                        tar.add(os.path.join(source_dir, key), arcname=key)
+
+    return stream
+
+
+@ray.remote
+class _PackActor:
+    """Actor wrapping around a packing job.
+
+    This actor is used for chunking the packed data into smaller chunks that
+    can be transferred via the object store more efficiently.
+
+    The actor will start packing the directory when initialized, and separate
+    chunks can be received by calling the remote ``next()`` task.
+
+    Args:
+        source_dir: Path to local directory to pack into tarfile.
+        files_stats: Dict of relative filenames mapping to a tuple of
+            (mtime, filesize). Only files that differ from these stats
+            will be packed.
+        chunk_size: Cut bytes stream into chunks of this size in bytes.
+    """
+
+    def __init__(
+        self,
+        source_dir: str,
+        files_stats: Optional[Dict[str, Tuple[float, int]]] = None,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,
+    ):
+        self.stream = _pack_dir(source_dir=source_dir, files_stats=files_stats)
+        self.chunk_size = chunk_size
+        self.iter = None
+
+    def get_full_data(self) -> bytes:
+        return self.stream.getvalue()
+
+    def _chunk_generator(self) -> Generator[None, bytes, None]:
+        self.stream.seek(0)
+        data = self.stream.read(self.chunk_size)
+        while data:
+            yield data
+            data = self.stream.read(self.chunk_size)
+
+    def next(self) -> Optional[bytes]:
+        if not self.iter:
+            self.iter = iter(self._chunk_generator())
+        try:
+            return next(self.iter)
+        except StopIteration:
+            return None
+
+
+def _iter_remote(actor: ray.ActorID) -> Generator[None, bytes, None]:
+    """Iterate over actor task and return as generator."""
+    while True:
+        try:
+            buffer = ray.get(actor.next.remote())
+            if buffer is None:
+                return
+            yield buffer
+        except RayTaskError:
+            return None
```

Comment:
Should we forward some type of error here?

Reply:
I've just removed the try-catch thing. Previously we used this to catch StopIteration from the remote actor but since we just return None here, we don't really need this.